server/tests/src/main/scala/sttp/tapir/server/tests/ServerWebSocketTests.scala

```diff
@@ -29,106 +30,143 @@ abstract class ServerWebSocketTests[F[_], S <: Streams[S], ROUTE](
-    ) { (backend, baseUri) =>
-      basicRequest
-        .response(asWebSocket { (ws: WebSocket[IO]) =>
-          for {
-            _ <- ws.sendText("""{"f":"apple"}""")
-            _ <- ws.sendText("""{"f":"orange"}""")
-            m1 <- ws.receiveText()
-            m2 <- ws.receiveText()
-          } yield List(m1, m2)
-        })
-        .get(baseUri.scheme("ws"))
-        .send(backend)
-        .map(_.body shouldBe Right(List("""{"f":"echo: apple"}""", """{"f":"echo: orange"}""")))
-    },
-    testServer(
-      endpoint.out(webSocketBody[String, CodecFormat.TextPlain, Option[String], CodecFormat.TextPlain](streams)),
-      "string server-terminated echo"
-    )((_: Unit) =>
-      pureResult(functionToPipe[String, Option[String]] {
-        case "end" => None
-        case msg   => Some(s"echo: $msg")
-      }.asRight[Unit])
-    ) { (backend, baseUri) =>
-      basicRequest
-        .response(asWebSocket { (ws: WebSocket[IO]) =>
-          for {
-            _ <- ws.sendText("test1")
-            _ <- ws.sendText("test2")
-            _ <- ws.sendText("end")
-            m1 <- ws.eitherClose(ws.receiveText())
-            m2 <- ws.eitherClose(ws.receiveText())
-            m3 <- ws.eitherClose(ws.receiveText())
-          } yield List(m1, m2, m3)
-        })
-        .get(baseUri.scheme("ws"))
-        .send(backend)
-        .map(
-          _.body.map(_.map(_.left.map(_.statusCode))) shouldBe Right(
-            List(Right("echo: test1"), Right("echo: test2"), Left(WebSocketFrame.close.statusCode))
+          .map(
+            _.body.map(_.map(_.left.map(_.statusCode))) shouldBe Right(
+              List(Right("echo: test1"), Right("echo: test2"), Left(WebSocketFrame.close.statusCode))
+            )
           )
-        )
-    },
-    testServer(
-      endpoint
-        .in(isWebSocket)
-        .errorOut(stringBody)
-        .out(stringWs),
-      "non web-socket request"
-    )(isWS => if (isWS) pureResult(stringEcho.asRight) else pureResult("Not a WS!".asLeft)) { (backend, baseUri) =>
-      basicRequest
-        .response(asString)
-        .get(baseUri.scheme("http"))
-        .send(backend)
-        .map(_.body shouldBe Left("Not a WS!"))
-    }
-  )
+      },
+      testServer(
+        endpoint
+          .in(isWebSocket)
+          .errorOut(stringBody)
+          .out(stringWs),
+        "non web-socket request"
+      )(isWS => if (isWS) pureResult(stringEcho.asRight) else pureResult("Not a WS!".asLeft)) { (backend, baseUri) =>
+        basicRequest
+          .response(asString)
+          .get(baseUri.scheme("http"))
+          .send(backend)
+          .map(_.body shouldBe Left("Not a WS!"))
+      },
+      testServer(
+        endpoint.out(stringWs),
+        "pong on ping"
+      )((_: Unit) => pureResult(stringEcho.asRight[Unit])) { (backend, baseUri) =>
+        basicRequest
+          .response(asWebSocket { (ws: WebSocket[IO]) =>
+            for {
+              _ <- ws.send(WebSocketFrame.ping)
```

Comment:
hm I'm not sure if all backends give acces to ping-pongs, so if this fails e.g. on akka-http we might need to parametrize the tests

Reply:
it passed